src/Microsoft.AspNet.Security.OAuthBearer/OAuthBearerAuthenticationHandler.cs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens;
+using System.Linq;
+using System.Runtime.ExceptionServices;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Security;
+using Microsoft.AspNet.Security.Infrastructure;
+using Microsoft.AspNet.Security.Notifications;
+using Microsoft.Framework.Logging;
+using Microsoft.IdentityModel.Protocols;
+
+namespace Microsoft.AspNet.Security.OAuthBearer
+{
+    internal class OAuthBearerAuthenticationHandler : AuthenticationHandler<OAuthBearerAuthenticationOptions>
+    {
+        private readonly ILogger _logger;
+        private OpenIdConnectConfiguration _configuration;
+
+        public OAuthBearerAuthenticationHandler(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        protected override AuthenticationTicket AuthenticateCore()
+        {
+            return AuthenticateCoreAsync().GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Searches the 'Authorization' header for a 'Bearer' token. If the 'Bearer' token is found, it is validated using <see cref="TokenValidationParameters"/> set in the options.
+        /// </summary>
+        /// <returns></returns>
+        protected override async Task<AuthenticationTicket> AuthenticateCoreAsync()
+        {
+            ExceptionDispatchInfo authFailedEx = null;
+            string token = null;
+            try
+            {
+                // Give application opportunity to find from a different location, adjust, or reject token
+                var messageReceivedNotification =
+                    new MessageReceivedNotification<HttpContext, OAuthBearerAuthenticationOptions>(Context, Options)
+                    {
+                        ProtocolMessage = Context,
+                    };
+
+                // notification can set the token
+                await Options.Notifications.MessageReceived(messageReceivedNotification);
+                if (messageReceivedNotification.HandledResponse)
+                {
+                    return messageReceivedNotification.AuthenticationTicket;
+                }
+
+                if (messageReceivedNotification.Skipped)
+                {
+                    return null;
+                }
+
+                string authorization = Request.Headers.Get("Authorization");
+                if (authorization.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
```

Comment:
@Praburaj: you don't have to add the token to the header, the message received notification doesn't assume any token or header.
@Tratcher: I like the idea of returning a token.
@PinpointTownes I wanted to leave options completely open. token may or may not be in a header we understand. that is why I choose 3 notifications.